File: Leetcode 0113 Path Sum II [easy] [tree]/Path Sum II [backtrack].java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> path = new ArrayList<>();
        
        findPathSum(root, sum, result, path);
        
        return result;
    }
    
    private void findPathSum(TreeNode cur, int sum, List<List<Integer>> result, List<Integer> path) {
        if (cur == null) {
            return;
        }
        
        if (cur.left == null && cur.right == null) {
            sum -= cur.val;
            if (sum == 0) {
                path.add(cur.val);
                result.add(new ArrayList<>(path));
                path.remove(path.size() - 1);
            }
            
            return;
        }
        
        path.add(cur.val);
        findPathSum(cur.left, sum - cur.val, result, path);
        findPathSum(cur.right, sum - cur.val, result, path);
        path.remove(path.size() - 1);
    }
}

File: Leetcode 0113 Path Sum II [easy] [tree]/Path Sum II [backtrack].js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number[][]}
 */
var pathSum = function(root, sum) {
    let result = [];
    let path = [];
    findPathSum(root, sum, result, path);
    return result;
};

let findPathSum = (cur, sum, result, path) => {
    if (cur === null) {
        return;
    }
    
    if (cur.left === null && cur.right === null) {
        sum = sum - cur.val;
        if (sum === 0) {
            path.push(cur.val);
            result.push(path.slice(0));
            path.pop();
        }
        return;
    }
    
    path.push(cur.val);
    findPathSum(cur.left, sum - cur.val, result, path);
    findPathSum(cur.right, sum - cur.val, result, path);
    path.pop();
}